### PR TITLE
Mark snippets and the bundler target's main file as having side effects

### LINKED ISSUE #972: There should be a way to disable `sideEffects: false`
## 🐛 Bug description

Custom JS snippets can contain top-level side-effects. Right now, wasm-pack generated package.json includes `sideEffects: false` which marks every file as side-effect free, leading to potential bugs after bundling.

#### 🤔 Expected Behavior

When JS snippet files are generated, they should be put into the `sideEffects` array, and only the JS file generated by wasm-bindgen should be assumed to be side-effect free since it's the only one wasm-bindgen has control over.

`package.json` allows to list potentially side-effect-y files as

```json
{
  "sideEffects": [
    "path/to/snippet.js"
  ]
}
```

This will avoid bugs for snippets with top-level side-effects, while it won't decrease optimization level for common scenarios where the module is used via its main entry point.

### PR DESCRIPTION
**Pull request by Liamolucko**
This is a less extreme version of #1208, which only marks snippets and the main file on the bundler target as having side effects instead of all files.

This means that the JS shim file which contains the vast majority of the JS code is still properly marked as having no side effects, allowing bundlers to get rid of things like unused `new TextEncoder` calls which could theoretically have side effects but don't.

Fixes #972
Fixes rustwasm/wasm-bindgen#3276

### CODE REVIEW

**Comment by cryptoquick** [comment]
I tried testing out your branch, but unfortunately I ran into this issue here: https://github.com/rustwasm/wasm-pack/issues/650#issuecomment-1436057987

It's unrelated to what's solved here, but if I can solve that, I can test if this resolves https://github.com/rustwasm/wasm-bindgen/issues/3306 and https://github.com/rustwasm/wasm-pack/issues/1230.

**Comment by Menci** [comment]
I think only `bundler` target's `sideEffects` should be set to the entry file. `web` target should keep `false`.

**Comment by Liamolucko** [comment]
> I think only `bundler` target's `sideEffects` should be set to the entry file. `web` target should keep `false`.

It's already effectively `false` for the entry file. On the web target, `sideEffects` is set to `["./snippets/*"]`, which means that only the files inside of `snippets` are treated as though they have side effects. Everything else, including the entry file, is treated as though `sideEffects` is set to false.

The reason why snippets are marked as having side effects is because they're arbitrary user code; we've got no idea what they do, and we don't want to assume that they have no side effects and cause them to break.

**Comment by martinxyz** [comment]
I've tested this branch (after a merge with master) and it fixed the issue for me.

**Comment by Sytten** [comment]
What's missing to move this forward? It's an annoying bug right now

**Comment by Liamolucko** [comment]
> What's missing to move this forward? It's an annoying bug right now

It's just waiting on a review from @drager.

**Review by drager — APPROVED** [review]
Thanks a lot!

**Comment by cryptoquick** [comment]
Thanks! 🙏

**Pull request merged**
